Sources/PackageGraph/TraitConfiguration.swift

```diff
@@ -11,18 +11,96 @@
 //===----------------------------------------------------------------------===//
 
 /// The trait configuration.
-public struct TraitConfiguration: Codable, Hashable {
-    /// The traits to enable for the package.
-    package var enabledTraits: Set<String>?
-
-    /// Enables all traits of the package.
-    package var enableAllTraits: Bool
+public enum TraitConfiguration: Codable, Hashable {
+    case enableAllTraits
+    case noConfiguration
+    case noEnabledTraits
+    case enabledTraits(Set<String>)
 
     public init(
         enabledTraits: Set<String>? = nil,
         enableAllTraits: Bool = false
     ) {
-        self.enabledTraits = enabledTraits
-        self.enableAllTraits = enableAllTraits
+        // If all traits are enabled, then no other checks are necessary.
+        guard !enableAllTraits else {
+            self = .enableAllTraits
+            return
+        }
+
+        if let enabledTraits {
+            if enabledTraits.isEmpty {
+                self = .noEnabledTraits
+            } else {
+                self = .enabledTraits(enabledTraits)
+            }
+        } else {
+            // Since enableAllTraits isn't enabled and there isn't a set of traits set,
+            // there is no configuration passed by the user.
+            self = .noConfiguration
+        }
+    }
+
+    /// Default instance of `TraitConfiguration`.
+    public static var `default`: TraitConfiguration {
+        .init(enabledTraits: nil)
     }
+
+    public var enabledTraits: Set<String>? {
+        switch self {
+        case .enabledTraits(let traits):
+            return traits
+        case .noConfiguration, .enableAllTraits:
```

Comment:
thought: it seems odd to me that the `.enableAllTraits` returns `nil`.  Should it return a set with all the traits instead?

Reply:
Since `.enableAllTraits` represents a flag that a user would pass in (and as such the `TraitConfiguration` is constructed before we begin parsing the `Manifest` which is where we'd have access to the entire list of traits that a manifest defines), that was the reasoning for it returning `nil` at this stage. Though since the `enum` now allows me to do specific checks for these cases where this information would be relevant, there may not be a need for this computed property - haven't fully refactored everything just yet.

Reply:
This has made me notice that it would probably make more sense to have `.noConfiguration` (or rather `.none` as the updated case name) return `["default"]` here. :)